kv/kvserver: TestLearnerOrJointConfigAdminRelocateRange failed

kv/kvserver.TestLearnerOrJointConfigAdminRelocateRange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8372389?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8372389?buildTab=artifacts#/) on master @ [529f7833f090a248e338fe6cf50ca35bc713ebdb](https://github.com/cockroachdb/cockroach/commits/529f7833f090a248e338fe6cf50ca35bc713ebdb):


```
=== RUN   TestLearnerOrJointConfigAdminRelocateRange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/33f576194975d55328ba8cc9ee712e65/logTestLearnerOrJointConfigAdminRelocateRange1745301299
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1292: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/991/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1292
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/991/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1317
        	Error:      	Received unexpected error:
        	            	cannot remove learner while snapshot is in flight
        	            	(1) forced error mark
        	            	  | "replication change in progress"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:1317
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6348
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6325
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:233
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (3) cannot remove learner while snapshot is in flight
        	            	Error types: (1) *markers.withMark (2) *errbase.opaqueWrapper (3) *errutil.leafError
        	Test:       	TestLearnerOrJointConfigAdminRelocateRange
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/33f576194975d55328ba8cc9ee712e65/logTestLearnerOrJointConfigAdminRelocateRange1745301299
--- FAIL: TestLearnerOrJointConfigAdminRelocateRange (2.44s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLearnerOrJointConfigAdminRelocateRange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23554

kv/kvserver.TestLearnerOrJointConfigAdminRelocateRange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8498890?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8498890?buildTab=artifacts#/) on master @ [69dd453d0e61e258f402c5751de310405743cd18](https://github.com/cockroachdb/cockroach/commits/69dd453d0e61e258f402c5751de310405743cd18):


```
=== RUN   TestLearnerOrJointConfigAdminRelocateRange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/2b93c386ff3e561b2014eb0644801d66/logTestLearnerOrJointConfigAdminRelocateRange3156250306
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1286: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/7550/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1286
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/7550/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1311
        	Error:      	Received unexpected error:
        	            	cannot remove learner while snapshot is in flight
        	            	(1) forced error mark
        	            	  | "replication change in progress"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:1317
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6348
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6325
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:233
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (3) cannot remove learner while snapshot is in flight
        	            	Error types: (1) *markers.withMark (2) *errbase.opaqueWrapper (3) *errutil.leafError
        	Test:       	TestLearnerOrJointConfigAdminRelocateRange
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/2b93c386ff3e561b2014eb0644801d66/logTestLearnerOrJointConfigAdminRelocateRange3156250306
--- FAIL: TestLearnerOrJointConfigAdminRelocateRange (5.40s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLearnerOrJointConfigAdminRelocateRange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestLearnerOrJointConfigAdminRelocateRange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8523755?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/8523755?buildTab=artifacts#/) on master @ [639decf058c78f0afe0dba613d60670628c92e81](https://github.com/cockroachdb/cockroach/commits/639decf058c78f0afe0dba613d60670628c92e81):


```
=== RUN   TestLearnerOrJointConfigAdminRelocateRange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/2b93c386ff3e561b2014eb0644801d66/logTestLearnerOrJointConfigAdminRelocateRange2603555480
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1286: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/8285/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1286
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/8285/execroot/com_github_cockroachdb_cockroach/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1311
        	Error:      	Received unexpected error:
        	            	cannot remove learner while snapshot is in flight
        	            	(1) forced error mark
        	            	  | "replication change in progress"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:1317
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6348
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6325
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:233
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1594
        	            	Wraps: (3) cannot remove learner while snapshot is in flight
        	            	Error types: (1) *markers.withMark (2) *errbase.opaqueWrapper (3) *errutil.leafError
        	Test:       	TestLearnerOrJointConfigAdminRelocateRange
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/2b93c386ff3e561b2014eb0644801d66/logTestLearnerOrJointConfigAdminRelocateRange2603555480
--- FAIL: TestLearnerOrJointConfigAdminRelocateRange (3.34s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLearnerOrJointConfigAdminRelocateRange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestLearnerOrJointConfigAdminRelocateRange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8573779?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8573779?buildTab=artifacts#/) on master @ [682ad1b0856591e366041b3008fa7eeefdb297c7](https://github.com/cockroachdb/cockroach/commits/682ad1b0856591e366041b3008fa7eeefdb297c7):


```
=== RUN   TestLearnerOrJointConfigAdminRelocateRange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/d351c1750c446bc1442d58db9c0ceef9/logTestLearnerOrJointConfigAdminRelocateRange4176336054
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1286: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/6876/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1286
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/6876/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1311
        	Error:      	Received unexpected error:
        	            	cannot remove learner while snapshot is in flight
        	            	(1) forced error mark
        	            	  | "replication change in progress"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:1317
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6348
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6325
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:233
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_arm64.s:1172
        	            	Wraps: (3) cannot remove learner while snapshot is in flight
        	            	Error types: (1) *markers.withMark (2) *errbase.opaqueWrapper (3) *errutil.leafError
        	Test:       	TestLearnerOrJointConfigAdminRelocateRange
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d351c1750c446bc1442d58db9c0ceef9/logTestLearnerOrJointConfigAdminRelocateRange4176336054
--- FAIL: TestLearnerOrJointConfigAdminRelocateRange (1.68s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLearnerOrJointConfigAdminRelocateRange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvserver.TestLearnerOrJointConfigAdminRelocateRange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8579555?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/8579555?buildTab=artifacts#/) on master @ [11b681286ebb667e61db058c953a7705a728e737](https://github.com/cockroachdb/cockroach/commits/11b681286ebb667e61db058c953a7705a728e737):


```
=== RUN   TestLearnerOrJointConfigAdminRelocateRange
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/15bcc1d4cfe2c01885bd956ef4f14c71/logTestLearnerOrJointConfigAdminRelocateRange1286212697
    test_log_scope.go:79: use -show-logs to present logs inline
    replica_learner_test.go:1485: 
        	Error Trace:	/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/10961/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1485
        	            				/home/roach/.cache/bazel/_bazel_roach/c5a4e7d36696d9cd970af2045211a7df/sandbox/processwrapper-sandbox/10961/execroot/com_github_cockroachdb_cockroach/bazel-out/aarch64-fastbuild/bin/pkg/kv/kvserver/kvserver_test_/kvserver_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/replica_learner_test.go:1510
        	Error:      	Received unexpected error:
        	            	cannot remove learner while snapshot is in flight
        	            	(1) forced error mark
        	            	  | "replication change in progress"
        	            	  | github.com/cockroachdb/errors/withstack/*withstack.withStack::
        	            	Wraps: (2)
        	            	  | (opaque error wrapper)
        	            	  | type name: github.com/cockroachdb/errors/withstack/*withstack.withStack
        	            	  | reportable 0:
        	            	  |
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver.init
        	            	  | 	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_command.go:1317
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6348
        	            	  | runtime.doInit
        	            	  | 	GOROOT/src/runtime/proc.go:6325
        	            	  | runtime.main
        	            	  | 	GOROOT/src/runtime/proc.go:233
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_arm64.s:1172
        	            	Wraps: (3) cannot remove learner while snapshot is in flight
        	            	Error types: (1) *markers.withMark (2) *errbase.opaqueWrapper (3) *errutil.leafError
        	Test:       	TestLearnerOrJointConfigAdminRelocateRange
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/15bcc1d4cfe2c01885bd956ef4f14c71/logTestLearnerOrJointConfigAdminRelocateRange1286212697
--- FAIL: TestLearnerOrJointConfigAdminRelocateRange (1.53s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLearnerOrJointConfigAdminRelocateRange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvserver/replica.go
+++ b/pkg/kv/kvserver/replica.go
@@ -2312,6 +2312,12 @@ func (r *Replica) GetLoadStatsForTesting() *load.ReplicaLoad {
 	return r.loadStats
 }
 
+// HasOutstandingLearnerSnapshotInFlightForTesting is for use only by tests to
+// gather whether there are in-flight snapshots to learner replcas.
+func (r *Replica) HasOutstandingLearnerSnapshotInFlightForTesting() bool {
+	return r.hasOutstandingLearnerSnapshotInFlight()
+}
+
 // ReadProtectedTimestampsForTesting is for use only by tests to read and update
 // the Replicas' cached protected timestamp state.
 func (r *Replica) ReadProtectedTimestampsForTesting(ctx context.Context) (err error) {

--- a/pkg/kv/kvserver/replica_learner_test.go
+++ b/pkg/kv/kvserver/replica_learner_test.go
@@ -1466,7 +1466,6 @@ func TestLearnerAndVoterOutgoingFollowerRead(t *testing.T) {
 
 func TestLearnerOrJointConfigAdminRelocateRange(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 95500, "flaky test")
 	defer log.Scope(t).Close(t)
 
 	ctx := context.Background()
@@ -1477,10 +1476,33 @@ func TestLearnerOrJointConfigAdminRelocateRange(t *testing.T) {
 	})
 	defer tc.Stopper().Stop(ctx)
 
+	tc.WaitForNodeLiveness(t)
+
 	scratchStartKey := tc.ScratchRange(t)
 	ltk.withStopAfterLearnerAtomic(func() {
 		_ = tc.AddVotersOrFatal(t, scratchStartKey, tc.Target(1))
 		_ = tc.AddVotersOrFatal(t, scratchStartKey, tc.Target(2))
+	})
+
+	{
+		// Ensure that the test starts with the expected number of learners:
+		// (n1,s1):1, (n2,s2):2 LEARNER, (n3,s3):3 LEARNER
+		desc := tc.LookupRangeOrFatal(t, scratchStartKey)
+		require.Len(t, desc.Replicas().LearnerDescriptors(), 2)
+	}
+
+	// Assert that initially there are no snapshots in flight to learners. This
+	// is an important precondition before testing AdminRelocateRange below, as
+	// AdminRelocateRange won't handle learners which have inflight snapshots and
+	// this test will flake.
+	testutils.SucceedsSoon(t, func() error {
+		desc := tc.LookupRangeOrFatal(t, scratchStartKey)
+		repl, err := tc.GetFirstStoreFromServer(t, 0).GetReplica(desc.RangeID)
+		require.NoError(t, err)
+		if repl.HasOutstandingLearnerSnapshotInFlightForTesting() {
+			return errors.Errorf("outstanding learner snapshot in flight %s", desc)
+		}
+		return nil
 	})
 
 	check := func(voterTargets []roachpb.ReplicationTarget) {

